lua/ttt2/extensions/math.lua

```diff
@@ -19,3 +19,48 @@ function math.ExponentialDecay(halflife, dt)
     -- ln(0.5) = -0.69..
     return exp((-0.69314718 / halflife) * dt)
 end
+
+---
+-- Gets the index of an item in the provided table, weighted according to the weights (derived from get_weight).
+-- @param table tbl The array of items to find a weighted item in.
+-- @param function get_weight Called as get_weight(item, index). Must return number.
+-- @return number
+-- @realm shared
+function math.random_weighted(tbl, get_weight)
+    -- There are several possible ways to get a weighted item. The most obvious is to simply include an item in
+    -- the table N times, where N is an integer proportional to the weight. This, however, requires maintaining
+    -- that table, which may be undesirable.
+    --
+    -- For a more friendly API, we instead compute the sum, generate a random number from 0 to that sum, then
+    -- take the first item whose weight prefix-sum is greater than that random number. This requires 2 passes
+    -- over the table, but enables processing on a normal table, with arbitrary weight storage.
+
+    -- Special case short arrays, because they may otherwise cause problems
+    if #tbl == 0 then
+        return nil
+    end
+    if #tbl == 1 then
+        return 1
+    end
+
+    -- first, compute the sum weight
+    local sum = 0
+    for k, v in pairs(tbl) do
+        sum = sum + get_weight(v, k)
+    end
+
+    -- get the random number
+    local rand = math.random(0, sum - 1)
```

Comment:
What if the sum is < 1 ?

Also this only returns integers, according to the docs, i think you meant to use math.Rand() 

https://wiki.facepunch.com/gmod/math.Rand

Reply:
`sum < 1` suggests either 1) zero or one table elements, which are explicitly checked for, or 2) weights that are `<= 0`, which are mostly nonsense. We need the `- 1` though because we need the upper bound to be exclusive to be correct.

And yes, integers are correct. The algorithm would work with reals, but they are not necessary.

Reply:
I think it should handle floats here, even if our special case does not use it, it is not clear from the outside and as a generic math function it should support those cases